### PR TITLE
CP-25570: connect to VNC console in qemu-upstream VMs using a Unix socket

### PR DESCRIPTION
**Pull request by mg12**
instead of using a network connection. This makes it easier to separate qemu-upstream in its own network namespace. A nice side-effect is that xenopsd knows the Unix path of the socket that was used when starting the qemu-upstream process and therefore QMP queries to qemu-upstream are not necessary anymore, simplifying the qemu-upstream backend.                                           

qemu-trad and PV VMs are not affected and continue to use a network connection for the VNC console.                                                                                           
                                                                                                            
This change depends on replacing the parameter for the VNC console from a network port to a Unix socket path when starting qemu-upstream, using the same matching vnc_socket_path used to connect to the console.

After this change, qemu-upstream will be started using 

    >  -vnc unix:/var/run/xen/vnc-<domid>,lock-key-sync=off
    >    A replacement for qemu-trad's -vncunused -vnc 127.0.0.1:1
                                                                                                    
instead of                                                                                                                                                               
                                                                                                                   
    >  -vnc 127.0.0.1:1,to=1000,lock-key-sync=off
    >    Equivalent to qemu-trad's -vncunused -vnc 127.0.0.1:1

This change depends on removing the corresponding vnc console parameter transformation in the qemu-upstream wrapper. The comment for the lock-key-sync=off option was moved from this wrapper.                                                                                                               
                                                                                                                                                                          
Signed-off-by: Marcus Granado <marcus.granado@citrix.com>   

----
Tests:

**Before change:**

_qemu-upstream:_ forkexecd: [ info|st45|0 ||forkexecd] qemu-dm-36[20159]: Arguments: 36 --syslog -d 36 -m 2048 -boot dc -serial pty -vcpus 2 -videoram 4 **-vncunused -vnc 127.0.0.1:1** -usb -usbdevice tablet -acpi -monitor null -device rtl8139,netdev=tapnet0,mac=1a:d6:a8:13:0d:dc,addr=4 -netdev tap,id=tapnet0,ifname=tap36.0,script=no,downscript=no

_qemu-trad:_ forkexecd: [ info|st45|0 ||forkexecd] qemu-dm-37[22415]: Arguments: qemu-dm-37 --syslog -d 37 -m 2048 -boot dc -serial pty -vcpus 2 -videoram 4 **-vncunused -vnc 127.0.0.1:1** -usb -usbdevice tablet -acpi -monitor null -net nic,vlan=0,macaddr=ca:22:79:17:4f:ce,model=rtl8139 -net tap,vlan=0,bridge=xenbr0,ifname=tap37.0

**After change:**

_qemu-upstream:_ forkexecd: [ info|st45|0 ||forkexecd] qemu-dm-16[22992]: Arguments: 16 --syslog -d 16 -m 2048 -boot dc -serial pty -vcpus 2 -videoram 4 **-vncunused -vnc unix:/var/run/xen/vnc-16,lock-key-sync=off** -usb -usbdevice tablet -acpi -monitor null -device rtl8139,netdev=tapnet0,mac=ea:c6:80:f2:cb:e3,addr=4 -netdev tap,id=tapnet0,ifname=tap16.0,script=no,downscript=no

_qemu-trad:_ forkexecd: [ info|st45|0 ||forkexecd] qemu-dm-19[32008]: Arguments: qemu-dm-19 --syslog -d 19 -m 2048 -boot dc -serial pty -vcpus 2 -videoram 4 **-vncunused -
vnc 127.0.0.1:1** -usb -usbdevice tablet -acpi -monitor null -net nic,vlan=0,macaddr=c2:be:98:51:e1:91,model=rtl8139 -net tap,vlan=0,bridge=xenbr0,ifname=tap19.0

### CODE REVIEW

File: xc/device.ml [hunk 1]
```diff
@@ -1671,7 +1692,10 @@ module Dm_Common = struct
         ]) info.disks in
 
     let restorefile = sprintf qemu_restore_path domid in
-    let disp_options, wait_for_port = cmdline_of_disp info in
+    let disp_options, wait_for_port = if domid_for_vnc
```
Comment (lindig) [inline xc/device.ml]: How does this work? `domid` is an argument that has no default. So how can `cmdline_of_disp info ~domid` and `cmdline_of_disp info` return even the same type?

File: xc/device.ml [hunk 2]
```diff
@@ -61,6 +61,14 @@ module Profile = struct
   let of_domid x = if is_upstream_qemu x then Qemu_upstream else Qemu_trad
 end
 
+(** Represent an IPC endpoint *)
+module Socket = struct
+  type port = Unix of string | Network of int
```
Comment (lindig) [inline xc/device.ml]: Maybe `Socket.t` would be a better type because `port` implies a number (at least for me).

File: xc/device.ml [hunk 3]
```diff
@@ -1612,7 +1622,27 @@ module Dm_Common = struct
     try Some (xs.Xs.read statepath)
     with _ -> None
 
-  let cmdline_of_disp info =
+  let vnc_opts_of ip_addr_opt auto port keymap ~domid =
+    let vnc_arg = match domid with
```
Comment (lindig) [inline xc/device.ml]:
```
let ip_addr = Opt.default "127.0.0.1" ip_addr_op in
let vnc_arg = match domid with
  None when auto -> Printf.sprintf "%s:1" ip_addr in
  None           -> Printf.sprintf "%s:%s" ip_addr port
  Some domid     -> Printf.sprintf "%s,lock-key-sync=off" (Socket.Unix.path (vnc_socket_path domid))
```

**Comment by rosslagerwall** [comment]
I think it would make sense while you're making these changes to remove this code from qemu-wrapper:

        if p == "-vncunused":
            del qemu_args[n]
            continue

And then ensure that this argument is _not_ passed to QEMU upstream.

**Comment by lindig** [comment]
Somewhat unrelated: I have tried to make the construction of the command line parameters more uniform and a little bit more obvious:

```
    List.concat
      [ [ "-d"; string_of_int domid
        ; "-m"; Int64.to_string (Int64.div info.memory 1024L)
        ; "-boot"; info.boot
        ]
      ; ( info.serial |> function None -> [] | Some x -> ["-serial";x])
      ; [ "-vcpus"; string_of_int info.vcpus]
      ; disp_options
      ; usb'
      ; List.concat disks'
      ; if info.acpi then [ "-acpi" ] else []
      ; if restore then [ "-loadvm"; restorefile ] else []
      ; (  info.pci_emulations
           |> List.map (fun pci -> ["-pciemulation"; pci])
           |> List.concat
        )
      ; if info.pci_passthrough then ["-priv"] else []
      ; (  info.extras
           |> List.map (function (k,None) -> ["-"^k] | (k,Some v) -> ["-"^k; v])
           |> List.concat
        )
      ; ( info.monitor  |> function None -> [] | Some x -> ["-monitor";x])
      ; ( info.parallel |> function None -> [] | Some x -> ["-parallel";x])
      ]
```

**Comment by lindig** [comment]
Can we progress the PR, @mg12?

File: xc/device.ml (hunk 2 above)
Comment (mg12) [reply xc/device.ml]: done on 4457f655e50b663954f616687886e9f0d20b5292

File: xc/device.ml (hunk 3 above)
Comment (mg12) [reply xc/device.ml]: done on 43bde950c06664b53adb86cda437cff0a4c618af

**Comment by mg12** [comment]
Pushed new commits with the suggested changes, after more devtests on pv, hvm-linux, hvm-windows qemu-trad/qemu-upstream vnc consoles.

**Pull request merged**